engine/executor/dm/unitholder.go

```diff
@@ -89,105 +68,142 @@ func newUnitHolderImpl(workerType lib.WorkerType, cfg *dmconfig.SubTaskConfig) *
 
 // Init implement UnitHolder.Init
 func (u *unitHolderImpl) Init(ctx context.Context) error {
+	u.processMu.Lock()
+	defer u.processMu.Unlock()
+
 	if err := u.unit.Init(ctx); err != nil {
 		return err
 	}
 
-	go u.unit.Process(u.ctx, u.resultCh)
+	runCtx, runCancel := context.WithCancel(context.Background())
+	u.fieldMu.Lock()
+	u.runCtx, u.runCancel = runCtx, runCancel
+	u.fieldMu.Unlock()
+
+	resultCh := make(chan pb.ProcessResult, 1)
+	u.processWg.Add(1)
+	go func() {
+		defer u.processWg.Done()
+		u.unit.Process(runCtx, resultCh)
+		u.fetchAndHandleResult(resultCh)
```

Comment:
Do you have any experience to reuse fetchResultAndUpdateStage?

Reply:
Not very realistic, they are owned by different objects（subtask/unitholder）. Maybe we can merge fetchResult into process later